### PR TITLE
Added non-inverse hyperbolic functions.

### PR DESCRIPTION
**Pull request by acjak**
The functions sinh and cosh are imported from the Eigen library,
similarly to other trigonometric functions.

### CODE REVIEW

**Comment by vrv** [comment]
Thanks for the contribution -- have you tested this out yourself already?

Also, I would recommend adding tests to cwise_ops_test.py so that we can validate its correctness ongoing.  Take a look at similar tests for ops in cwise_ops.h

**Comment by drpngx** [comment]
Following from http://stackoverflow.com/questions/42314317/eigen-functor-undeclared-when-building-tensorflow.

Jenkins, test this please.

**Comment by drpngx** [comment]
The error appears to be that `cosh` (at least) is not defined for `Eigen::half`, only double and long double.

**Comment by acjak** [comment]
@vrv I did not initially test it. Turns out it was not as simple (refer to the link supplied by @drpngx ). I now got it to compile, but the sinh and cosh functions are not available in Python after TF import.
I think this PR should be dropped. I will continue working on it and make another PR eventually.

**Comment by drpngx** [comment]
Exporting to python should be relatively easy once the op is registered.
I'll outline the steps

**Comment by drpngx** [comment]
OK, there is some documentation here, but it's for adding an op outside of the code base.
https://www.tensorflow.org/versions/r0.10/how_tos/adding_an_op/

The steps are: build the kernel, as you did. You should `REGISTER_KERNEL` for the various devices and data types that you want to support. Second, create an op and call `REGISTER_OP`. That is the wrapper "declaration" of the op that will cause some python code to be generated. BTW, you'll also need a gradient for best results. The op should now be imported like the other math ops, supposing you put them in the same place as `sin`. Finally, you should declare the ops as part of the public API (which will require review), [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_ops.py#L50). I would suggest exporting as `tf.sinh` rather than `tf.nn.sinh` even though we have `tf.nn.tanh`.

**Comment by acjak** [comment]
Ok, thanks a lot for the pointer. I've been looking a little more into it when I've had the time. All in all I have now:

- Added the sinh and cosh templates to `cwise_ops.h`.
- Added the files (kernel files) `cwise_op_sinh.cc` and `cwise_op_cosh.cc`.
- Corrected the data types in the input arguments to be just float and double.
- Added `REGISTER_OP` for both 'Cosh' and 'Sinh' in `math_ops.cc`.
- Added op definitions of 'Cosh' and 'Sinh' in ops.pbtxt (I don't know if this is necessary).

It compiles now although I have not tried with CUDA and openCL yet. I can also import tensorflow in Python, but `tf.sinh` and `tf.cosh` are still not available.

HOWEVER: I can import them with `from tensorflow.python.ops.gen_math_ops import cosh sinh`. And they work. I just can't figure out why they are not picked up along with the other functions from gen_math_ops.py, where I can find cosh and sinh along with the other trigonometric functions.

**Comment by drpngx** [comment]
You need to add them [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_ops.py#L50) as well.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Oh, since this is an API change, you'll need to modify the API [goldens](https://www.github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/api/golden/tensorflow.pbtxt#L1589)

**Comment by drpngx** [comment]
This doesn't link on Linux GPU (at least).

```
cwise_op_cosh.cc:(.text._ZN10tensorflow7UnaryOpIN5Eigen9GpuDeviceENS_7functor4coshIfEEE7ComputeEPNS_15OpKernelContextE[_ZN10tensorflow7UnaryOpIN5Eigen9GpuDeviceENS_7functor4coshIfEEE7ComputeEPNS_15OpKernelContextE]+0x143): undefined reference to `tensorflow::functor::UnaryFunctor<Eigen::GpuDevice, tensorflow::functor::cosh<float> >::operator()(Eigen::GpuDevice const&, Eigen::TensorMap<Eigen::Tensor<float, 1, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<Eigen::Tensor<float const, 1, 1, long>, 16, Eigen::MakePointer>)'
```

**Comment by acjak** [comment]
Yeah, that look like the same problem as for CPU, `cosh` and `sinh` are not defined for the data types in the kernels.

**Comment by acjak** [comment]
Or rather, it's the Eigen::half that's the problem again.

**Comment by acjak** [comment]
That link is dead for me.

**Comment by drpngx** [comment]
Never mind this.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
```
ERROR: /workspace/tensorflow/core/kernels/BUILD:2098:1: C++ compilation of rule '//tensorflow/core/kernels:cwise_op' failed: crosstool_wrapper_driver_is_not_gcc failed: error executing command 
  (cd /var/lib/jenkins/workspace/tensorflow-pull-requests-gpu/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/workspace && \
  exec env - \
    LD_LIBRARY_PATH=/usr/local/cuda/extras/CUPTI/lib64:/usr/local/nvidia/lib:/usr/local/nvidia/lib64 \
    PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin \
  external/local_config_cuda/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc -U_FORTIFY_SOURCE '-D_FORTIFY_SOURCE=1' -fstack-protector -fPIE -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 -DNDEBUG -ffunction-sections -fdata-sections '-std=c++11' -MD -MF bazel-out/local_linux-opt/bin/tensorflow/core/kernels/_objs/cwise_op/tensorflow/core/kernels/cwise_op_cosh.d '-frandom-seed=bazel-out/local_linux-opt/bin/tensorflow/core/kernels/_objs/cwise_op/tensorflow/core/kernels/cwise_op_cosh.o' -DEIGEN_MPL2_ONLY -DTENSORFLOW_USE_JEMALLOC -iquote . -iquote bazel-out/local_linux-opt/genfiles -iquote external/bazel_tools -iquote bazel-out/local_linux-opt/genfiles/external/bazel_tools -iquote external/eigen_archive -iquote bazel-out/local_linux-opt/genfiles/external/eigen_archive -iquote external/local_config_sycl -iquote bazel-out/local_linux-opt/genfiles/external/local_config_sycl -iquote external/jemalloc -iquote bazel-out/local_linux-opt/genfiles/external/jemalloc -iquote external/protobuf -iquote bazel-out/local_linux-opt/genfiles/external/protobuf -iquote external/gif_archive -iquote bazel-out/local_linux-opt/genfiles/external/gif_archive -iquote external/jpeg -iquote bazel-out/local_linux-opt/genfiles/external/jpeg -iquote external/com_googlesource_code_re2 -iquote bazel-out/local_linux-opt/genfiles/external/com_googlesource_code_re2 -iquote external/farmhash_archive -iquote bazel-out/local_linux-opt/genfiles/external/farmhash_archive -iquote external/highwayhash -iquote bazel-out/local_linux-opt/genfiles/external/highwayhash -iquote external/png_archive -iquote bazel-out/local_linux-opt/genfiles/external/png_archive -iquote external/zlib_archive -iquote bazel-out/local_linux-opt/genfiles/external/zlib_archive -iquote external/local_config_cuda -iquote bazel-out/local_linux-opt/genfiles/external/local_config_cuda -iquote external/curl -iquote bazel-out/local_linux-opt/genfiles/external/curl -iquote external/boringssl -iquote bazel-out/local_linux-opt/genfiles/external/boringssl -iquote external/jsoncpp_git -iquote bazel-out/local_linux-opt/genfiles/external/jsoncpp_git -isystem external/bazel_tools/tools/cpp/gcc3 -isystem external/eigen_archive -isystem bazel-out/local_linux-opt/genfiles/external/eigen_archive -isystem external/jemalloc/include -isystem bazel-out/local_linux-opt/genfiles/external/jemalloc/include -isystem external/protobuf/src -isystem bazel-out/local_linux-opt/genfiles/external/protobuf/src -isystem external/gif_archive/lib -isystem bazel-out/local_linux-opt/genfiles/external/gif_archive/lib -isystem external/farmhash_archive/src -isystem bazel-out/local_linux-opt/genfiles/external/farmhash_archive/src -isystem external/png_archive -isystem bazel-out/local_linux-opt/genfiles/external/png_archive -isystem external/zlib_archive -isystem bazel-out/local_linux-opt/genfiles/external/zlib_archive -isystem external/local_config_cuda/cuda/include -isystem bazel-out/local_linux-opt/genfiles/external/local_config_cuda/cuda/include -isystem external/local_config_cuda/cuda -isystem bazel-out/local_linux-opt/genfiles/external/local_config_cuda/cuda -isystem external/curl/include -isystem bazel-out/local_linux-opt/genfiles/external/curl/include -isystem external/boringssl/src/include -isystem bazel-out/local_linux-opt/genfiles/external/boringssl/src/include -isystem external/jsoncpp_git/include -isystem bazel-out/local_linux-opt/genfiles/external/jsoncpp_git/include -DEIGEN_AVOID_STL_ARRAY -Iexternal/gemmlowp -Wno-sign-compare -fno-exceptions '-DGOOGLE_CUDA=1' -msse3 -pthread '-DGOOGLE_CUDA=1' -no-canonical-prefixes -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -fno-canonical-system-headers -c tensorflow/core/kernels/cwise_op_cosh.cc -o bazel-out/local_linux-opt/bin/tensorflow/core/kernels/_objs/cwise_op/tensorflow/core/kernels/cwise_op_cosh.o): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 1.
tensorflow/core/kernels/cwise_op_cosh.cc:33:61: error: macro "REGISTER3" requires 7 arguments, but only 6 given
 REGISTER3(UnaryOp, GPU, "Cosh", functor::cosh, float, double);
                                                             ^
tensorflow/core/kernels/cwise_op_cosh.cc:33:1: error: 'REGISTER3' does not name a type
 REGISTER3(UnaryOp, GPU, "Cosh", functor::cosh, float, double);
 ^
```

**Comment by drpngx** [comment]
Jenkins, test this please.

It would probably be faster if you tested this on your end first.

**Comment by acjak** [comment]
I compiled it locally, but it's on a laptop so I don't have CUDA support. This error seems to be something new: 

> Error checking context: 'no permission to read from '/var/lib/jenkins/workspace/tensorflow-pull-requests-makefile/tensorflow/contrib/makefile/downloads/eigen/Eigen/src/Core/arch/CUDA/Complex.h''.

**Comment by drpngx** [comment]
The GPU error is:

```
cwise_op_cosh.cc:(.text._ZN10tensorflow7UnaryOpIN5Eigen9GpuDeviceENS_7functor4coshIfEEE7ComputeEPNS_15OpKernelContextE[_ZN10tensorflow7UnaryOpIN5Eigen9GpuDeviceENS_7functor4coshIfEEE7ComputeEPNS_15OpKernelContextE]+0x163): undefined reference to `tensorflow::functor::UnaryFunctor<Eigen::GpuDevice, tensorflow::functor::cosh<float> >::operator()(Eigen::GpuDevice const&, Eigen::TensorMap<Eigen::Tensor<float, 1, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<Eigen::Tensor<float const, 1, 1, long>, 16, Eigen::MakePointer>)'
```

You can still install the toolchain on your machine (or a docker) and build for GPU.

**Comment by drpngx** [comment]
Any luck with that? BTW, we also need to register the derivatives (`cosh <-> sinh`) for best results.

**Comment by drpngx** [comment]
Jenkins, test this please.

Testing if makefile is a transient error.

**Comment by drpngx** [comment]
Please add the corresponding cuda files, aping the [cos](https://www.github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/cwise_op_gpu_cos.cu.cc#L18) version.

**Comment by drpngx** [comment]
Any luck with this?

**Comment by dandelionmane** [comment]
Adding the stalled label since it's been two weeks without a response. @acjak if you can't get to this soon we'll close it to keep our list of open PRs manageable, but you'd be welcome to open a new PR with the same changes when you have time to focus on it.

**Comment by martinwicke** [comment]
Closing now. Feel free to reopen any time with identical changes (and add the cuda files so it'll link for GPU)
